client/app/pages/queries/utils/useVisualizationTabHandler.js

```diff
@@ -0,0 +1,35 @@
+import { useState, useEffect, useMemo } from "react";
+import { first, orderBy } from "lodash";
+import { $location, $rootScope } from "@/services/ng";
+
+function updateUrlHash(...args) {
+  $location.hash(...args);
+  $rootScope.$applyAsync();
+}
+
+export default function useVisualizationTabHandler(visualizations) {
+  const firstVisualization = useMemo(() => first(orderBy(visualizations, ["id"])), [visualizations]);
+  const [selectedTab, setSelectedTab] = useState(+$location.hash() || firstVisualization.id);
+
+  useEffect(() => {
+    const newHashValue = selectedTab !== firstVisualization.id ? `${selectedTab}` : null;
+    if ($location.hash() !== newHashValue) {
+      updateUrlHash(newHashValue);
+    }
+  }, [firstVisualization.id, selectedTab]);
+
+  useEffect(() => {
+    const unwatch = $rootScope.$watch(
+      () => $location.hash(),
+      () => {
+        const expectedHashValue = selectedTab !== firstVisualization.id ? `${selectedTab}` : null;
+        if ($location.hash() !== expectedHashValue) {
+          setSelectedTab(+$location.hash() || firstVisualization.id);
+        }
+      },
+    );
+    return unwatch;
+  }, [firstVisualization.id, selectedTab]);
```

Comment:
WDYT about merging these two `useEffect` calls? They have the same dependencies and their logic is quite related.

Reply:
Done :+1:. LMK if you had more simplification in mind for that part

Reply:
I'd also inline `updateUrlHash` function, but it's up to you. Code looks good :+1:

Reply:
Hi, this feature can be add to dashboard page? Thank you!